File: Applied Course/4.Problem Solving/5.Problems on Trees/23.Search in a Binary Search Tree.py
# https://leetcode.com/problems/search-in-a-binary-search-tree/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# Logic-1
class Solution:
    def searchBST(self, root: TreeNode, val: int) -> TreeNode:
        if not root or root.val==val:
            return root
        if root.val>val:
            return self.searchBST(root.left,val)
        if root.val<val:
            return self.searchBST(root.right,val)
# Logic-2
class Solution:
    def searchBST(self, root: TreeNode, val: int) -> TreeNode:
        while root!=None and root.val!=val:
            if root.val>val:
                root=root.left
            else:
                root=root.right
        return root

File: Applied Course/4.Problem Solving/5.Problems on Trees/23.Search in a Binary Search Tree.cpp
#include<bits/stdc++.h>
using namespace std;
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// Logic-1
class Solution {
public:
    TreeNode* searchBST(TreeNode* root, int val) {
        if(!root || root->val==val)
            return root;
        if(root->val>val)
            return searchBST(root->left,val);
        return searchBST(root->right,val);
    }
};
// Logic-2
class Solution {
public:
    TreeNode* searchBST(TreeNode* root, int val) {
        while(root!=NULL && root->val!=val)
            if(root->val>val)
                root=root->left;
            else
                root=root->right;
        return root;
    }
};
int main()
{
    // https://leetcode.com/problems/search-in-a-binary-search-tree/
    return 0;
}